Plugins/NativeEngine/Source/NativeEngine.cpp

```diff
@@ -742,21 +744,31 @@ namespace Babylon
         const Napi::Function onError = info[3].As<Napi::Function>();
         ProgramData* program = new ProgramData{};
         arcana::make_task(arcana::threadpool_scheduler, *m_cancellationSource,
-            [this, vertexSource, fragmentSource, program, cancellationSource{m_cancellationSource}]() -> bgfx::ProgramHandle
+            [this, vertexSource, fragmentSource, program, cancellationSource{m_cancellationSource}]() -> std::pair<bgfx::ProgramHandle, std::unordered_map<std::string, uint32_t>>
             {
                 return CreateProgramInternal(vertexSource, fragmentSource, *program);
             })
-            .then(m_runtimeScheduler, *m_cancellationSource, [program, onSuccessRef{Napi::Persistent(onSuccess)}, onErrorRef{Napi::Persistent(onError)}, cancellationSource{m_cancellationSource}](arcana::expected<bgfx::ProgramHandle, std::exception_ptr> result)
+            .then(m_runtimeScheduler, *m_cancellationSource, [program, onSuccessRef{Napi::Persistent(onSuccess)}, onErrorRef{Napi::Persistent(onError)}, cancellationSource{m_cancellationSource}](arcana::expected<std::pair<bgfx::ProgramHandle, std::unordered_map<std::string, uint32_t>>, std::exception_ptr> result)
                 {
-                if (result.has_error())
-                {
-                    onErrorRef.Call({});
-                }
-                else
-                {
-                    program->Handle = result.value();
-                    onSuccessRef.Call({});
-                } });
+                    if (result.has_error())
+                    {
+                        try
+                        {
+                            std::rethrow_exception(result.error());
+                        }
+                        catch (const std::exception& ex)
+                        {
+                            onErrorRef.Call({Napi::Error::New(onErrorRef.Env(), ex.what()).Value()});
+                        }
+                    }
+                    else
+                    {
+                        std::pair<bgfx::ProgramHandle, std::unordered_map<std::string, uint32_t>> programInfo = result.value();
```

Comment:
This will result in a copy.

```suggestion
                        auto& programInfo = result.value();
```

Reply:
I pushed the changes we discussed (using `std::move` out of the result value) and hopefully that addresses this issue.